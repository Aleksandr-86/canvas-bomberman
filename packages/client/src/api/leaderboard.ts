import { TEAM_NAME } from './../features/constants'
import axios from 'axios'
import { API_URL } from '../features/constants'
import { PlayerStats } from '../store/reducers/leaderboardSlice'

export type LeadersRequest = {
  ratingFieldName: string
  cursor: number
  limit: number
}

const options = {
  withCredentials: true,
  headers: {
    'content-type': 'application/json',
  },
}

const sendLeaderboardNewLeaderRequest = async (data: PlayerStats) => {
  const requestParams = {
    ratingFieldName: 'score',
    teamName: TEAM_NAME,
  }

  const response = await axios.post<PlayerStats>(
    `${API_URL}/leaderboard`,
    { data: data, ...requestParams },
    { ...options }
  )

  return response
}

const getTeamLeaderboard = async (data: LeadersRequest) => {
  const response = await axios.post(
    `${API_URL}/leaderboard/${TEAM_NAME}`,
    data,
    { ...options }
  )

  return response
}

export const LeaderboardAPI = {
  sendLeaderboardNewLeaderRequest,
  getTeamLeaderboard,
}
