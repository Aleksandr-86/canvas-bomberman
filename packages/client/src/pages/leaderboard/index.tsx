import { Link } from 'react-router-dom'

import { Table } from '../../components/table/table'
import { Button } from '../../components/button/button'

import styles from './leaderboard.module.css'

// Заглушка
const tableData = [
  {
    id: 1,
    place: 1,
    name: 'Bionic',
    score: 27500,
    games: 10,
    totalTime: 61,
  },
  {
    id: 2,
    place: 2,
    name: 'Chrome Road',
    score: 15150,
    games: 5,
    totalTime: 36,
  },
  {
    id: 3,
    place: 3,
    name: 'Bladewatch',
    score: 13500,
    games: 4,
    totalTime: 30,
  },
  {
    id: 4,
    place: 4,
    name: 'Marshmallow',
    score: 12100,
    games: 3,
    totalTime: 23,
  },
  {
    id: 5,
    place: 5,
    name: 'Arctic Realm',
    score: 9800,
    games: 3,
    totalTime: 19,
  },
]

export const Leaderboard = () => {
  return (
    <div className={styles.leaderboard}>
      <div className={styles.leaderboard__wrapper}>
        <Table caption="Таблица лидеров" userStats={tableData} />
        <Link to="/">
          <Button class={styles.leaderboard__button} title="На главную" />
        </Link>

        <img
          className={styles['leaderboard__hero-img']}
          src="src/assets/images/hero.png"
          alt="hero-img"
          draggable="false"
        />
      </div>
    </div>
  )
}
