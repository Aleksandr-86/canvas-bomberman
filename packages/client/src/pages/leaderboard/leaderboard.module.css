.leaderboard {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 1920px;
  height: 1000px;
  user-select: none;
}

.container {
  display: flex;
  flex-direction: column;
  gap: 50px;
  align-items: center;
  width: 1250px;
}

.tableCaption {
  font-size: 40px;
  text-transform: uppercase;
}

.buttonWrapper {
  display: flex;
  flex-direction: row;
  gap: 20px;
}

.heroImg {
  position: absolute;
  top: 780px;
  right: 150px;
  width: 253px;
  height: 191px;
}
