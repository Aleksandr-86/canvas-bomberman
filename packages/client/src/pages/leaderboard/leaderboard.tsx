import { Link } from 'react-router-dom'

import { useAppDispatch, useAppSelector } from '../../store/hooks'
import {
  getLeaderboardState,
  getUser,
  isLeaderboardLoading,
} from '../../store/selectors'
import { NavigationBar } from '../../components/navigationBar/navigationBar'
import { LeaderboardTable } from '../../components/leaderboardTable/leaderboardTable'
import heroImg from '../../assets/images/hero.png'

import baseStyles from '../../app/app.module.css'
import styles from './leaderboard.module.css'
import { Button } from '../../components/button/button'
import { nextPage, previousPage } from '../../store/reducers/leaderboardSlice'
import { leaderboardThunks } from '../../store/actions/leaderboardThunks'
import { useEffect } from 'react'
import { ActivityIndicator } from '../../components/activityIndicator/activityIndicator'
import { me } from '../../store/actions/userActions'

export const Leaderboard = () => {
  const userState = useAppSelector(getUser)
  const leaderboardState = useAppSelector(getLeaderboardState)
  const isLoading = useAppSelector(isLeaderboardLoading)
  const dispatch = useAppDispatch()

  useEffect(() => {
    dispatch(leaderboardThunks.setLeadersStatsState())
  }, [])

  const leaderboard = (
    <div className={styles.container}>
      <div className={styles.tableCaption}>Таблица лидеров</div>
      <LeaderboardTable state={leaderboardState} />

      <div className={styles.buttonWrapper}>
        <Button onClick={() => dispatch(previousPage())}>&lt; </Button>
        <Button onClick={() => dispatch(nextPage())}>&gt;</Button>
        <Link className={baseStyles.linkButton} to="/">
          На главную
        </Link>
      </div>
      <img
        className={styles.heroImg}
        src={heroImg}
        alt="hero-img"
        draggable="false"
      />
    </div>
  )

  const navBar = userState.isAuth ? <NavigationBar /> : null
  const pageContent = isLoading ? <ActivityIndicator /> : leaderboard

  return (
    <div className={styles.leaderboard}>
      {navBar}

      {pageContent}
    </div>
  )
}
