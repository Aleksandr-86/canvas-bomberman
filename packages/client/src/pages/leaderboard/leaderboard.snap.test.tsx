import * as renderer from 'react-test-renderer'
import { Leaderboard } from './leaderboard'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

const mockStore = configureStore([])
const store = mockStore({
  user: {
    isAuth: true,
    user: {
      displayName: `mockDisplayName`,
      login: `mockLogin`,
      firstName: `mockFirstName`,
      secondName: `mockSecondName`,
      email: `mockEmail`,
      phone: 8900000000,
    },
  },
  leaderboard: {
    isLoading: false,
    stats: [
      {
        id: 1,
        place: 1,
        name: `Alex`,
        score: 811,
        games: 1,
      },
    ],
    page: 0,
  },
})

describe(`<Leaderboard/> snapshot test cases`, () => {
  test(`<Leaderboard/> should render <div className=leaderboard> parent with child elements (with store)`, () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter initialEntries={[`/leaderboard`]}>
            <Leaderboard />
          </MemoryRouter>
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
