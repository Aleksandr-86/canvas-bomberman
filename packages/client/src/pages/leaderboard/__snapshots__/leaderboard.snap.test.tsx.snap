// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Leaderboard/> snapshot test cases <Leaderboard/> should render <div className=leaderboard> parent with child elements (with store) 1`] = `
<div
  className="leaderboard"
>
  <nav
    className="navigationBar"
  >
    <img
      alt="user-avatar"
      draggable="false"
    />
    <a
      className="link"
      href="/profile"
      onClick={[Function]}
    >
      mockDisplayName
    </a>
    <a
      className="link"
      href="/"
      onClick={[Function]}
    >
      Выйти
    </a>
  </nav>
  <div
    className="container"
  >
    <div
      className="tableCaption"
    >
      Таблица лидеров
    </div>
    <table
      className="table"
    >
      <thead>
        <tr>
          <th
            className="head ascending"
            onClick={[Function]}
          >
            Место
          </th>
          <th
            className="head"
            onClick={[Function]}
          >
            Имя игрока
          </th>
          <th
            className="head"
            onClick={[Function]}
          >
            Очки
          </th>
          <th
            className="head"
            onClick={[Function]}
          >
            Сыграно игр
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="row"
        >
          <td
            className="columnPlace"
          >
            1
          </td>
          <td
            className="columnName"
          >
            Alex
          </td>
          <td
            className="columnScore"
          >
            811
          </td>
          <td
            className="columnGames"
          >
            1
          </td>
        </tr>
      </tbody>
    </table>
    <div
      className="buttonWrapper"
    >
      <button
        className="button"
        onClick={[Function]}
      >
        &lt; 
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        &gt;
      </button>
      <a
        className="linkButton"
        href="/"
        onClick={[Function]}
      >
        На главную
      </a>
    </div>
    <img
      alt="hero-img"
      className="heroImg"
      draggable="false"
    />
  </div>
</div>
`;
